### PR TITLE
feat: add qc only workflow

### PR DESCRIPTION
**Pull request by rannick**
Added: qc only workflow

### Review and tests: 
- [x] Tests pass
- [x] Code review APJ
- [x] New code is executed and covered by tests, and test approve

Balsamic config qc command:
```
balsamic config qc_panel --case-id <test> --balsamic-cache <balsamic_cache> --tumor <S1_R_1.fastq.gz> --normal <S2_R_1.fastq.gz>  --case-id <test> --panel-bed <panel.bed> --analysis-dir <test>
```

Help:
```
Usage: balsamic config qc [OPTIONS]

Options:
  --case-id TEXT                  Sample id that is used for reporting,
                                  naming the analysis jobs, and analysis path
                                  [required]
  --quality-trim / --no-quality-trim
                                  Trim low quality reads in fastq  [default:
                                  quality-trim]
  --adapter-trim / --no-adapter-trim
                                  Trim adapters from reads in fastq  [default:
                                  adapter-trim]
  -p, --panel-bed PATH            Panel bed file for variant calling.
  --balsamic-cache PATH           Path to BALSAMIC cache  [required]
  --container-version [develop|master|8.2.8]
                                  Container for BALSAMIC version to download
                                  [default: 8.2.8]
  --analysis-dir PATH             Root analysis path to store analysis logs
                                  and results.           The final path will
                                  be analysis-dir/sample-id  [required]
  -t, --tumor PATH                Fastq files for tumor sample.  [required]
  -n, --normal PATH               Fastq files for normal sample.
  --umi / --no-umi                UMI processing steps for samples with UMI
                                  tags.For WGS cases, UMI is always disabled.
                                  [default: umi]
  --umi-trim-length INTEGER       Trim N bases from reads in fastq  [default:
                                  5]
  --tumor-sample-name TEXT        Tumor sample name
  --normal-sample-name TEXT       Normal sample name
  -g, --genome-version [hg19|hg38|canfam3]
                                  Genome version to prepare reference. Path to
                                  genomewill be <outdir>/genome_version
  --help                          Show this message and exit.

```

The following rules will be included in the workflow: ['snakemake_rules/quality_control/fastp.rule', 'snakemake_rules/quality_control/fastqc.rule', 'snakemake_rules/quality_control/multiqc.rule', 'snakemake_rules/variant_calling/mergetype_tumor.rule', 'snakemake_rules/quality_control/picard.rule', 'snakemake_rules/quality_control/sambamba_depth.rule', 'snakemake_rules/quality_control/mosdepth.rule', 'snakemake_rules/align/bwa_mem.rule']

Output: 
```
Job stats:
job                                      count    min threads    max threads
-------------------------------------  -------  -------------  -------------
all                                          1              1              1
bwa_mem                                      2              1              1
fastp                                        2              1              1
fastp_umi                                    2              1              1
fastqc                                       2              1              1
mergeBam_tumor                               1              1              1
mosdepth_coverage                            2              1              1
multiqc                                      1              1              1
picard_CollectAlignmentSummaryMetrics        2              1              1
picard_CollectHsMetrics                      2              1              1
picard_CollectInsertSizeMetrics              2              1              1
picard_markduplicates                        2              1              1
sambamba_exon_depth                          2              1              1
sambamba_panel_depth                         2              1              1
total                                       25              1              1
```

### CODE REVIEW

**Comment by ashwini06** [comment]
@rannick Let's discuss this PR tomorrow before I start adding lots of comments 😄

File: BALSAMIC/workflows/QC.smk [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+quality_control_results = [result_dir + "qc/" + "multiqc_report.html"]
+
+if 'delivery' in config:
+    wildcard_dict = {"sample": list(config["samples"].keys())+["tumor", "normal"],
+                     "case_name": config["analysis"]["case_id"],
+                     "allow_missing": True
+                     }
+
+    if config['analysis']["analysis_type"] in ["paired", "single"]:
+        wildcard_dict.update({"bedchrom": config["panel"]["chrom"] if "panel" in config else [],
```
Comment (ashwini06) [inline BALSAMIC/workflows/QC.smk:115]: might not be used by any qc rules.. please cross-check

File: BALSAMIC/workflows/QC.smk [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+# picarddup flag
+picarddup = get_picard_mrkdup(config)
+
+# parse parameters as constants to workflows
+params = BalsamicWorkflowConfig.parse_obj(WORKFLOW_PARAMS)
+
+# Capture kit name
+if config["analysis"]["sequencing_type"] != "wgs":
+    capture_kit = os.path.split(config["panel"]["capture_kit"])[1]
```
Comment (ashwini06) [inline BALSAMIC/workflows/QC.smk:54]:
```suggestion

```
I don't think we need this  capture_kit info to run any qc rules

**Comment by ashwini06** [comment]
@rannick any information about which bedfile will be used for running dog genomes? Is canfam3 is the genome build requested by the customer?

File: .github/workflows/black_linter.yml [hunk 3]
```diff
@@ -11,4 +11,4 @@ jobs:
       - uses: psf/black@stable
         with:
           options: "--check --verbose"
-          version: "21.7b0"
+          version: "22.3.0"
```
[the remark this answers is not in the document]
Comment (rannick) [reply .github/workflows/black_linter.yml:14]: Addressed in 7813839

File: BALSAMIC/workflows/QC.smk [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+from BALSAMIC.utils.cli import (write_json, check_executable, generate_h5)
+
+from BALSAMIC.utils.models import VarCallerFilter, BalsamicWorkflowConfig
+
+from BALSAMIC.utils.rule import (get_rule_output, get_result_dir,
+                                 get_sample_type, get_picard_mrkdup, get_script_path,
+                                 get_threads, get_sequencing_type, get_capture_kit)
+
+from BALSAMIC.constants.common import (RULE_DIRECTORY, VCFANNO_TOML, MUTATION_TYPE);
+from BALSAMIC.constants.workflow_params import WORKFLOW_PARAMS
```
[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/workflows/QC.smk:23]: https://github.com/Clinical-Genomics/BALSAMIC/blob/7813839207f46b2c06e4ed8d87edbc504fca6a9d/BALSAMIC/snakemake_rules/align/bwa_mem.rule#L20

File: BALSAMIC/workflows/QC.smk [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+# Add reference assembly if not defined for backward compatibility
+if 'genome_version' not in config["reference"]:
+    GENOME_VERSION = 'hg19' ## if hg19 convention works, replace accordingly
+    LOG.info('Genome version was not found in config. Setting it to %s', GENOME_VERSION)
+
+# Add normal sample if analysis is paired
+germline_call_samples = ["tumor"]
+if config['panel']:
+    germline_call_samples.append("normal")
```
[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/workflows/QC.smk]: Indeed not needed; addressed in f9c3800

File: BALSAMIC/commands/config/qc.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    type=int,
+    help="Trim N bases from reads in fastq",
+)
+@click.option(
+    "--umiworkflow/--no-umiworkflow",
+    default=False,
+    show_default=True,
+    is_flag=True,
+    help="Enable running UMI workflow",
+)
```
[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/commands/config/qc.py]: Actually, multiqc needs a umiworkflow parameter: https://github.com/Clinical-Genomics/BALSAMIC/blob/8914b315ba8cf4486871a46707801dccc648bc71/BALSAMIC/snakemake_rules/quality_control/multiqc.rule#L59

[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/commands/config/qc.py]: Set to False

File: BALSAMIC/workflows/QC.smk (hunk 1 above)
Comment (rannick) [reply BALSAMIC/workflows/QC.smk:107]: Not used for qc rules

File: BALSAMIC/workflows/QC.smk (hunk 2 above)
Comment (rannick) [reply BALSAMIC/workflows/QC.smk:54]: capture_kit is used in mosdepth and multiqc

**Comment by rannick** [comment]
canine_exome_16.1 will be the bed file

File: BALSAMIC/commands/config/qc.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    is_flag=True,
+    help="Trim adapters from reads in fastq",
+)
+@click.option(
+    "-p",
+    "--panel-bed",
+    type=click.Path(exists=True, resolve_path=True),
+    required=False,
+    help="Panel bed file for variant calling.",
+)
```
[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/commands/config/qc.py:56]: We will need

**Review by ashwini06 — COMMENTED** [review]
Now the only thing I am wondering is that qc_only workflow will be panel specific, as the `config qc` command line requires a mandatory panel-bed option. Thinking, if we can modify slightly the `rules_to_include` in qc.smk to work along with wgs cases as well.

File: BALSAMIC/utils/models.py [hunk 8]
```diff
@@ -196,10 +196,11 @@ class AnalysisModel(BaseModel):
     Attributes:
 
         case_id : Field(required); string case identifier
-        analysis_type : Field(required); string literal [single, paired, pon]
+        analysis_type : Field(required); string literal [single, paired, pon, qc]
             single : if only tumor samples are provided
             paired : if both tumor and normal samples are provided
             pon : panel of normal analysis
+            qc : QC analysis only
```
Comment (ashwini06) [inline BALSAMIC/utils/models.py:203]:
```suggestion
            qc_panel : QC analysis only
```

File: BALSAMIC/workflows/QC.smk [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        panel_bed = get_capture_kit(config)
+    run:
+        import datetime
+        import shutil
+
+        from BALSAMIC.utils.qc_metrics import get_qc_metrics_json
+
+        # Save QC metrics to a JSON file
+        try:
+            qc_metrics_summary = get_qc_metrics_json(params.result_dir, params.sequencing_type, params.panel_bed)
```
Comment (ashwini06) [inline BALSAMIC/workflows/QC.smk:154]:
```suggestion
            qc_metrics_summary = get_qc_metrics_json(params.result_dir, params.sequencing_type, params.panel_bed)
```
We don't know any qc check criteria for dog exome panel, we might need to add the dog panel here https://github.com/Clinical-Genomics/BALSAMIC/blob/454c0c450d37c4b73b55e0daf0ff3ed102ef919d/BALSAMIC/constants/quality_check_reporting.py#L53. @ivadym Any inputs here on how to handle qc metrics here?

**Review by ashwini06 — CHANGES_REQUESTED** [review]
@rannick : `qc` is not changed `qc_panel` in some places of code. Can you cross check once again please.

File: BALSAMIC/commands/config/qc.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+from BALSAMIC.utils.models import BalsamicConfigModel
+
+
+LOG = logging.getLogger(__name__)
+
+
+@click.command(
+    "qc_panel",
+    short_help="Create a sample config file from input sample data for QC analysis",
```
[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/commands/config/qc.py]: in [70182b2](https://github.com/Clinical-Genomics/BALSAMIC/pull/847/commits/70182b2a404159d1537b309dbc9fdbbac064f53f)

File: BALSAMIC/commands/config/qc.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    help="Fastq files for normal sample.",
+)
+@click.option(
+    "--umi/--no-umi",
+    default=True,
+    show_default=True,
+    is_flag=True,
+    help=(
+        "UMI processing steps for samples with UMI tags."
+        "For WGS cases, UMI is always disabled."
```
[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/commands/config/qc.py]: in [70182b2](https://github.com/Clinical-Genomics/BALSAMIC/pull/847/commits/70182b2a404159d1537b309dbc9fdbbac064f53f)

File: BALSAMIC/utils/cli.py [hunk 12]
```diff
@@ -277,6 +277,8 @@ def get_snakefile(analysis_type, reference_genome="hg19"):
 
     if analysis_type == "pon":
         snakefile = Path(p, "workflows", "PON.smk")
+    if "qc" in analysis_type:
```
[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/utils/cli.py]: in [70182b2](https://github.com/Clinical-Genomics/BALSAMIC/pull/847/commits/70182b2a404159d1537b309dbc9fdbbac064f53f)

File: BALSAMIC/utils/models.py [hunk 13]
```diff
@@ -215,7 +216,7 @@ class AnalysisModel(BaseModel):
 
     Raises:
         ValueError:
-            When analysis_type is set to any value other than [single, paired, qc, pon]
+            When analysis_type is set to any value other than [single, paired, pon, qc]
```
[the remark this answers is not in the document]
Comment (rannick) [reply BALSAMIC/utils/models.py]: in [70182b2](https://github.com/Clinical-Genomics/BALSAMIC/pull/847/commits/70182b2a404159d1537b309dbc9fdbbac064f53f)

File: BALSAMIC/utils/models.py (hunk 8 above)
Comment (rannick) [reply BALSAMIC/utils/models.py]: resolved in [70182b2](https://github.com/Clinical-Genomics/BALSAMIC/pull/847/commits/70182b2a404159d1537b309dbc9fdbbac064f53f)

File: BALSAMIC/workflows/QC.smk (hunk 9 above)
Comment (rannick) [reply BALSAMIC/workflows/QC.smk]: Updated qc metrics in [97a8738](https://github.com/Clinical-Genomics/BALSAMIC/pull/847/commits/97a8738dc8b58091631367e210ae508d89b2c321). Defaults thresholds for targeted qc metrics are used

**Review by ashwini06 — APPROVED** [review]
Great job 🎆  With successful test runs,  this PR is ready to merge 👍🏼

**Comment by rannick** [comment]
Test results available in /home/proj/development/cancer/canfam3_testruns/small

**Pull request merged**